feat: Add Unit Tests for Mempool Functionality

eugypalu opened the pull request:
Resolves: #1355 

# Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing

# What is the new behavior?

mempool unit tests

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

tcoratger commented:
@eugypalu Nice addition to our test suite. I would have two small remarks before merging:
- I think we will wait to merge https://github.com/kkrt-labs/kakarot-rpc/pull/1352 first because there will be a small change in the provider but normally it will just require to modify very slightly the way you call the pool which should be in the ethclient and no longer in the ethprovider (but it is not a big change for your tests).
- On the other hand, I know that it is rather recommended to separate the tests to have unit test as much as possible and limit the scope of each test but in our case it is a little different. Each time we need to start Katana and the docker container so it still takes a little time. So I would be in favor of grouping some of your tests together, for example:
   - If I'm not mistaken, your `test_mempool_pooled_transaction_hashes` test checks the hashes of transactions added to the pool against the expected. So we could make these assertions directly at the end of your `test_mempool_add_transactions` test. This would add additional assertions but it would limit the number of tests.
   - I think you can do the same thing to group several tests together as soon as possible.

eugypalu commented:
> On the other hand, I know that it is rather recommended to separate the tests to have unit test as much as possible and limit the scope of each test but in our case it is a little different. Each time we need to start Katana and the docker container so it still takes a little time. So I would be in favor of grouping some of your tests together, for example:
> * If I'm not mistaken, your `test_mempool_pooled_transaction_hashes` test checks the hashes of transactions added to the pool against the expected. So we could make these assertions directly at the end of your `test_mempool_add_transactions` test. This would add additional assertions but it would limit the number of tests.
> * I think you can do the same thing to group several tests together as soon as possible.

Fixed, thanks

tcoratger commented:
Nice, lgtm, pending https://github.com/kkrt-labs/kakarot-rpc/pull/1352 to be merged

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+        // Get the Ethereum provider
+        let eth_provider = katana.eth_provider();
+
+        // Get the chain ID
+        let chain_id = eth_provider.chain_id().await.unwrap_or_default().unwrap_or_default().to();
```
can be moved out of the for loop

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+            input: Bytes::default(),
+            max_fee_per_gas: 875_000_000,
+            max_priority_fee_per_gas: 0,
+            access_list: Default::default(),
```
I think you can do `..Default::default()` for `input` and `access_list`.

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+
+        // Create a signed transaction
+        let transaction_signed = TransactionSigned::from_transaction_and_signature(transaction, signature);
+        // Recover the signer from the transaction
+        let signer = transaction_signed.recover_signer().ok_or(SignatureError::Recovery)?;
```
I think you can do `katana.eoa().address()` instead of recovering the signer. Also, since this will be the same each loop, you can just do it outside the for loop

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+    assert_eq!(mempool_size.queued, 0);
+    assert_eq!(mempool_size.total, 0);
+
+    // all_transactions function tests
+    // Check if the first pending transaction hash matches
+    assert_eq!(*all_transactions.pending[0].hash(), transaction_signed.hash());
+    // Ensure only one pending transaction is present
+    assert_eq!(all_transactions.pending.len(), 1);
+    // Ensure no queued transactions are present
+    assert_eq!(all_transactions.queued.len(), 0);
```
can we move this above the `remove_transaction` test? It's a little confusing when reading the test

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+    assert_eq!(mempool_size.total, transaction_number);
+
+    // pending_transactions function tests
+    // Check if the first pending transaction hash matches
+    assert_eq!(hashes[0].hash(), &signed_transactions[0].hash());
+    // Check if the second pending transaction hash matches
+    assert_eq!(hashes[1].hash(), &signed_transactions[1].hash());
+    // Ensure the number of pending transactions matches the expected count
+    assert_eq!(hashes.len(), transaction_number);
+
+    // get_transactions_by_sender function tests
+    // Ensure only one transaction is returned
+    assert_eq!(sender_transactions.len(), transaction_number);
+    // Check if the returned transaction hash matches
+    assert_eq!(*sender_transactions[0].hash(), signed_transactions[0].hash());
+    assert_eq!(*sender_transactions[1].hash(), signed_transactions[1].hash());
```
can't we replace all this by retrieving expected_hashes in a vec and received hashes in another vec?

greged93 replied:
also I think it's nicer when the assert follows the actual function call directly, it's easier to follow the test.
So something like
```rust
    // Get pending transactions
    let hashes = eth_provider.mempool().unwrap().pending_transactions();
    // Check if the first pending transaction hash matches
    assert_eq!(hashes[0].hash(), &signed_transactions[0].hash());
    // Check if the second pending transaction hash matches
    assert_eq!(hashes[1].hash(), &signed_transactions[1].hash());
    // Ensure the number of pending transactions matches the expected count
    assert_eq!(hashes.len(), transaction_number);```

eugypalu replied:
fixed on all the tests, thanks

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+    // Check total transactions
+    assert_eq!(mempool_size.total, 2);
+
+    // pooled_transaction_hashes function tests
+    // Check if the first signed transaction hash is present
+    assert!(hashes.contains(&signed_transactions[0].hash()));
+    // Check if the second signed transaction hash is present
+    assert!(hashes.contains(&signed_transactions[1].hash()));
+    // Ensure the hashes are not empty
+    assert!(!hashes.is_empty());
+
+    // pooled_transaction_hashes_max function test
+    // Check if at least one signed transaction hash is present
+    assert!(hashes_max.contains(&signed_transactions[0].hash()) || hashes_max.contains(&signed_transactions[1].hash()));
+    // Ensure the number of hashes matches the limit
+    assert_eq!(hashes_max.len(), hashes_max_number);
+    // Ensure the hashes are not empty
+    assert!(!hashes_max.is_empty());
```
same as above

greged93 commented on tests/tests/mempool.rs:
```diff
@@ -0,0 +1,396 @@
+#[rstest]
+#[awt]
+#[tokio::test(flavor = "multi_thread")]
+async fn test_mempool_add_transaction_and_subscribe(#[future] katana: Katana, _setup: ()) {
```
we aren't really testing the subscription here right?

eugypalu replied:
No, it wasn’t testing the subscription. I removed it, I don’t think it was useful. What do you think?

greged93 replied:
I think it makes sense to remove if you weren't actually testing the subscription yes

tcoratger commented:
@eugypalu Here I think that you will have some conflicts to solve after the merge of https://github.com/kkrt-labs/kakarot-rpc/pull/1352#event-14148899607. Can you rebase? The main thing is that now you have to call `mempool()` and `send_raw_transaction()` from the `eth_client` and not from the `eth_provider` anymore.

eugypalu commented:
> @eugypalu Here I think that you will have some conflicts to solve after the merge of [#1352 (comment)](https://github.com/kkrt-labs/kakarot-rpc/pull/1352#event-14148899607). Can you rebase? The main thing is that now you have to call `mempool()` and `send_raw_transaction()` from the `eth_client` and not from the `eth_provider` anymore.

Hey @tcoratger, I've rebased and fixed the conflicts 👍
